Validate path parameters match parameters

kevinmeredith opened the pull request:
Addresses https://github.com/swagger-api/swagger-parser/issues/189.

Note - it's a **WIP**. I will `git rebase` when I believe it's ready for review. I'm using the PR to review the work that I've done via the PR "Files Changed" UI.

The intention of this PR is to fail, i.e. return `null` for the parsed `Swagger` object, if any `path`'s value, e.g. `/pets/{id}`, has an unaccounted for path parameter. Note that, for a path parameter of `{id}`, the global, path, and operation-specific path parameters would be retrieved, and then checked whether they, collectively, contain `{id}`.

I've added quite of bit of tests, but I'm going to add more to account for path parameters with global and path-specific path parameters.

kevinmeredith commented:
> I've added quite of bit of tests, but I'm going to add more to account for path parameters with global and path-specific path parameters.

Done.

Reviewers, please check out this PR.

kevinmeredith commented:
@webron pointed out that my PR is wrong to include top-level `parameters`: https://github.com/swagger-api/swagger-parser/issues/189#issuecomment-183083193.

When I've updated the PR accordingly per Ron's reply, I will post here.

kevinmeredith commented:
I just updated this PR to exclude `global` parameters, per Ron's comment.

However, a single test failed:

```
Failed tests: 
  SwaggerReaderTest.testIssueSwaggerUi1792:114 NullPointer
```

This test uses `thingy.json`, whose ` `"/things/{id}"` path has an unaccounted for `id` path parameter. In other words, the `id` path parameter is not specified in either the path or operation `parameters`.

Should I just add `id` as either type of parameter, and then re-run the test to verify it succeeds?

fehguy commented:
Let's leave your test--I will verify a fix today and put it in when merging.  Thanks for the contrib!

kevinmeredith commented:
> Let's leave your test

To confirm that I understand - does this mean that I should leave the PR as-is, i.e. you'll verify the fix and fix the broken test, @fehguy?

Thank you

fehguy commented:
That's correct--I'll review the broken test and make sure it's either passing with another fix, or the test is disabled if it doesn't make sense.

kevinmeredith commented:
Hi @fehguy - could you please let me know if you've had a chance to review this PR?

Thanks

kevinmeredith commented:
Any updates, please? 

Thanks

fehguy commented:
Sorry for the delay.  I've been traveling and this takes a bit of time--I will review today.

fehguy commented on modules/swagger-parser/src/main/java/io/swagger/parser/Swagger20Parser.java:
```diff
@@ -129,6 +133,23 @@ private Swagger convertToSwagger(String data) throws IOException {
+                    if(validPaths.isPresent()) {
+                        final Boolean validResult = validPaths.get();
+                        if(validResult.equals(Boolean.FALSE)) {
+                            return null;
```
We can't abort the parse if there are parameters missing.  The `SwaggerDeserializer` is responsible for keeping track of messages and returning a potentially invalid spec if there are problems.

fehguy commented on modules/swagger-parser/src/main/java/io/swagger/parser/util/SwaggerDeserializer.java:
```diff
@@ -174,6 +176,176 @@ else if(!ROOT_KEYS.contains(key)) {
         return output;
     }
 
+    public static Optional<Boolean> allPathParametersAccountForExtractedPathParameters(final Swagger swagger) {
```
Why use `Optional<Boolean>` instead of `boolean`?  We only have true/false for this operation.

fehguy commented:
OK @kevinmeredith I've gone though this and there are some challenges.

First, conceptually the parser's job is to _parse_ the spec, and provide messages back if it fails.

Next, the parser should not return null unless it's unreadable, or of a different version than the parser expects.  Null means "I don't know what to do" which is different from "this is a swagger 2.0 definition but there are issues".

Finally, and most importantly, we _cannot_ validate parameters unless we either (a) do full resolution in the validation process (such as, resolving remote references) or (b) check structural validation issues _after_ we have run the Resolver.

That said, there are some valuable checks in here, but I don't see how we can use the PR without some major changes.

I suggest working in the `SwaggerParser` and adding logic to this method:

```
SwaggerDeserializationResult readWithInfo(String location, List<AuthorizationValue> auths, boolean resolve)
```

_If_ `resolve = true`, you can then check for existing parameters _then_ add them to the messages in the `SwaggerDeserializationResult` as feedback rather than an error.

kevinmeredith commented:
Thanks for the feedback. I expect to reply to it, i.e. ask questions about particular changes, within 1 week or so.

kevinmeredith commented:
Sorry for the delay in addressing your comments, Tony. Thanks for your time in reviewing it.

In the event that I get time to issue another PR, I will reference this one.

Thanks

fehguy commented:
no worries, I'm trying to get a release out--we can always reopen or start a new ticket when it makes sense.
